File: client2/src/game/systems/Checkpoint.ts
import { System } from '../../ecs/System'
import { SystemFlag } from '../flags'
import { NeverAspect } from '../../ecs/Aspect'
import { hasComponent, check } from '../utils/collision'
import { Player } from '../components/Player'
import { Checkpoint } from '../components/Checkpoint'

export class CheckpointSystem extends System {
  static _id = 'Checkpoint'
  _id = CheckpointSystem._id

  static _flag = SystemFlag.Checkpoint
  _flag = CheckpointSystem._flag

  static _aspect = new NeverAspect()
  _aspect = CheckpointSystem._aspect

  beginContact = (a: Fixture, b: Fixture, contact: Contact) => {
    const result = check(a, b, [hasComponent(Player), hasComponent(Checkpoint)])

    if (result) {
      const player = result[0].as(Player)
      const checkpoint = result[1].as(Checkpoint)

      if (player && checkpoint) {
        checkpoint.visited = true
        player.checkpoint = Math.max(checkpoint.index)
      }
    }
  }
}

File: archive/client/src/game/systems/checkpoint.lua
local Aspect = require 'src.ecs.aspect'
local System = require 'src.ecs.system'
local Checkpoint = require 'src.game.components.checkpoint'
local Player = require 'src.game.components.player'
local collision = require 'src.game.utils.collision'

local CheckpointSystem = System:new('checkpoint', Aspect.always())

local function checkCollision(a, b)
  return a.meta.type == 'player' and b.meta.type == 'checkpoint'
end

local function updateCheckpoint(player, checkpoint)
  player.checkpoint = math.max(player.checkpoint, checkpoint.index)
  checkpoint.visted = true
end

function CheckpointSystem:beginContact(a, b, contact)
  a = a:getUserData().entity
  b = b:getUserData().entity

  if checkCollision(a, b) then
    updateCheckpoint(a:as(Player), b:as(Checkpoint))
  elseif checkCollision(b, a) then
    updateCheckpoint(b:as(Player), a:as(Checkpoint))
  end
end

return CheckpointSystem
